leapp/reporting/__init__.py

```diff
@@ -1,106 +1,249 @@
+        raise NotImplementedError
 
-        inhibitor - This report entry will inhibit the upgrade process
-    """
+    def apply(self, report):
+        """Add report entry to the report dict based on provided path"""
+        _add(report, self.path, self.value, self.kind)
+
+
+class Title(BasePrimitive):
+    """Report title"""
+    @property
+    def path(self):
+        return ('title',)
+
+
+class Summary(BasePrimitive):
+    """Report summary"""
+    @property
+    def path(self):
+        return ('summary',)
+
+
+class Severity(BasePrimitive):
+    """Report severity"""
+    def __init__(self, value=None):
+        if not isinstance(value, str):
+            raise TypeError('Value of "{}" must be a string'.format(self.__class__.__name__))
+        severities = ('info', 'low', 'medium', 'high')
+        if value not in severities:
+            raise ValueError('Severity should be one of: "{}"'.format(' '.join(severities)))
+        self._value = value
+
+    @property
+    def path(self):
+        return ('severity',)
+
+
+class Audience(BasePrimitive):
+    """Target audience of the report"""
+    def __init__(self, value=None):
+        if not isinstance(value, str):
+            raise TypeError('Value of "{}" must be a string'.format(self.__class__.__name__))
+        audiences = ('sysadmin', 'developer')
+        if value not in audiences:
+            raise ValueError('Audience should be one of: "{}"'.format(' '.join(audiences)))
+        self._value = value
+
+    @property
+    def path(self):
+        return ('audience',)
+
+
+class Flags(BasePrimitive):
+    """Report flags"""
+    def __init__(self, value=None):
+        if not isinstance(value, list):
+            raise TypeError('Value of "{}" must be a list'.format(self.__class__.__name__))
+        self._value = value
 
+    @property
+    def path(self):
+        return ('flags',)
 
-def report(title=None, detail=None, renderers=None, audience=None, flags=None, severity=None):
+
+class Tags(BasePrimitive):
+    """Report tags"""
+
+    ACCESSIBILITY = 'accessibility'
+    COMMUNICATION = 'communication'
+    DRIVERS = 'drivers'
+    FILESYSTEMS = 'filesystems'
+    HIGH_AVAILABILITY = 'high availability'
+    SECURITY = 'security'
+    SELINUX = 'selinux'
+    TOOLS = 'tools'
+
+    def __init__(self, value=None):
+        if not isinstance(value, list):
+            raise TypeError('Value of "{}" must be a list'.format(self.__class__.__name__))
+        self._value = value
+
+    @property
+    def path(self):
+        return ('tags',)
+
+
+class ExternalLink(BasePrimitive):
+    """External link report detail field"""
+    def __init__(self, url=None, title=None):
+        if not all(isinstance(v, str) for v in (url, title)):
+            raise TypeError('Values "url" and "title" of "{}" must be a string'.format(self.__class__.__name__))
+        self._value = {'url': url, 'title': title}
+
+    @property
+    def kind(self):
```

Comment:
It's not so intuitive to guess what kind 0 and kind 1 means. Can you rename this to something like is_somebooleanprop (is_list? is_dict?), so that the name would actually correspond to the value?

Reply:
fixed